Migrate notes data store to wp data

joshuatf opened the pull request:
Fixes #3380

Moves the notes data store from wc-api to wp data under the `data` package. 

* Migrates selectors, actions, and resolvers to the new data store.  Caches all notes under `state.notes` and individual queries reference note IDs from `state.noteQueries`.
* Updates references to these methods in components.
* Makes some updates around checking for requesting/resolving status and note updating status (now stored in individual note).  /cc @octaedro for your review on this piece to make sure things work as expected 😄 
* Moves the notices for dismissal into the components.

### Screenshots
<img width="527" alt="Screen Shot 2020-08-06 at 8 26 25 AM" src="https://user-images.githubusercontent.com/10561050/89531647-3d177f00-d7f9-11ea-8a65-78217c54ea40.png">


### Detailed test instructions:

1. Check all places that retrieve notes (activity panel inbox / historical data / unread indicators / store alerts / profile wizard).
1. Test dismissing individual and all notes.
1. Try to undo both the above dismissals using the undo button in the notice.
1. Make sure notes loading status appears correctly when resolving.

joshuatf committed:
Add initial notes data store using wp data

joshuatf committed:
Cache individual note queries

joshuatf committed:
Update notes undo endpoint to batch update endpoint

joshuatf committed:
Allow batch updates of notes

joshuatf committed:
Add requesting state

joshuatf committed:
Update components to use new note store

joshuatf committed:
Remove old notes store from wc-api

joshuatf committed:
Get note IDs from json and validate data in REST API

joshuatf committed:
Add an updating state to each note

joshuatf committed:
Throw errors in actions so they can be caught by client

joshuatf committed:
Add store reducer tests

joshuatf commented on src/API/Notes.php:
```diff
@@ -327,15 +327,25 @@ public function get_requested_updates( $request ) {
 	 * @return WP_REST_Request|WP_Error
 	 */
-	public function undoremove_items( $request ) {
+	public function batch_update_items( $request ) {
```
I realized this endpoint is already set up to batch update, which could be pretty useful.

Not sure if we need to add a deprecated method for `undoremove_items` if we go forward with this update.

timmyc replied:
> Not sure if we need to add a deprecated method for undoremove_items if we go forward with this update.

I think that is the safe thing to do given prior feedback on matters like this.

joshuatf replied:
Added deprecated method and notice.

octaedro reviewed (commented):
Hi @joshuatf, you did an awesome job here! The code looks great, I just left a comment and wanted to ask you if we should remove:
* `getUndoDismissRequesting` from [client/homescreen/layout.js](https://github.com/woocommerce/woocommerce-admin/blob/update/note-data-store/client/homescreen/layout.js#L167) too? It looks like we should use `isBatchUpdating` instead.
* and the `notes` references from [client/wc-api/wc-api-spec.js](https://github.com/woocommerce/woocommerce-admin/blob/update/note-data-store/client/wc-api/wc-api-spec.js#L7).

octaedro commented on client/header/activity-panel/panels/inbox/index.js:
```diff
@@ -54,7 +48,7 @@ const renderNotes = ( {
 
 	return notesArray.map( ( note ) => {
-		if ( isDismissUndoRequesting === note.id ) {
+		if ( note.isUpdating ) {
```
When we are updating a note (when undoing the dismission) `note.isUpdating` is `undefined` but should be `true` to be able to show the placeholder in its place.

joshuatf committed:
Handle PR feedback

joshuatf commented:
Thanks for that feedback, @octaedro and great catches!  Handled all your feedback and the deprecated method in https://github.com/woocommerce/woocommerce-admin/pull/4943/commits/b56ee5118214bf8a6f212a4a2c69a65c77af2778.

This is ready for re-review.

octaedro commented:
I'm experiencing some trouble in the `Home` screen / `Dashboard`. It's throwing a couple of console errors when I dismiss all the notes and later `Undo` the dismissal.
Also, probably related, the tests are failing on `WC_Tests_API_Admin_Notes::test_undo_all_notes_delete`.

joshuatf committed:
Fix failing tests

joshuatf committed:
Don't set inbox panel visibility while batch updating

joshuatf commented:
Thanks again for reviewing, @octaedro 😄 

Fixed failing tests in https://github.com/woocommerce/woocommerce-admin/pull/4943/commits/4ec45c4678fbb2234779263bb8ce5a04c9020562 and fixed the frontend infinite loop in https://github.com/woocommerce/woocommerce-admin/pull/4943/commits/c59c0371e436f5e12895c2959e72fa47bbcc697e.

Also gave this PR a rebase.  Could you give it another look when you have a moment?

octaedro reviewed (approved):
Great job @joshuatf! Everything is working well here and code looks good :shipit:

joshuatf committed:
Move woocommerce block to external dependencies

psealock commented:
@joshuatf I'm getting a couple warnings on `main` that seem related to this PR. Can you reproduce?

![Screen Shot 2020-08-14 at 2 09 54 PM](https://user-images.githubusercontent.com/1922453/90206514-f0a2f380-de37-11ea-948a-917966895743.png)

joshuatf commented:
I can, thanks for the heads up, @psealock.  PR opened [here](https://github.com/woocommerce/woocommerce-admin/pull/4990) to address that and cleaning up those constants.
